File: backend/models/Wall.js
const mongoose = require("mongoose");

const STATUS_ENUM = [
    "OPEN",
    "CLOSED",
    "UNDER_MAINTAINANCE",
    "PERMANENTLY_CLOSED",
];
const DIFFICULTY_ENUM = ["UNKNOWN", "BEGINNER", "INTERMEDIATE", "ADVANCED", "EXPERT"];

const baseWallTransform = (doc, ret) => {
    ret.id = ret._id;
    delete ret._id;
    delete ret.__v;
    return ret;
};

const wallSchema = new mongoose.Schema(
    {
        name: {
            type: String,
            required: true,
            trim: true,
        },
        description: {
            type: String,
            trim: true,
            maxLength: [
                1000,
                "Description cannot be more than 1000 characters",
            ],
        },
        location: {
            type: {
                type: String,
                enum: ["Point"],
                default: "Point",
            },
            coordinates: {
                type: [Number], // [Longitude, Latitude]
                required: true,
            },
            address: String,
        },
        rating: {
            type: Number,
            default: 0,
        },
        difficulty: {
            type: String,
            enum: DIFFICULTY_ENUM,
            required: true,
        },
        status: {
            type: String,
            enum: STATUS_ENUM,
            default: "OPEN",
        },
        sessions: [
            {
                type: mongoose.Schema.Types.ObjectId,
                ref: "ClimbingSession",
            },
        ],
    },
    {
        timestamps: true,
        discriminatorKey: "wallType",
        toJSON: {
            transform: baseWallTransform,
            virtuals: true,
        },
        toObject: { virtuals: true },
    },
);

wallSchema.index({ location: "2dsphere" });
wallSchema.index({ name: "text" });

wallSchema.methods.editWall = function (updates) {};

wallSchema.methods.computeRating = async function () {
    const ClimbingSession = mongoose.model("ClimbingSession");

    const sessions = await ClimbingSession.find({
        wall_id: this._id,
        review_id: { $ne: null },
    }).populate("review_id", "rating");

    if (!sessions.length) {
        this.rating = 0;
        await this.save();
        return this.rating;
    }

    const totalRating = sessions.reduce((sum, session) => {
        return sum + (session.review_id?.rating || 0);
    }, 0);

    this.rating = totalRating / sessions.length;
    await this.save();

    return this.rating;
};

// Virtual for counting sessions related to this wall
wallSchema.virtual("totalSessions").get(function () {
    return this._totalSessions || 0;
});

wallSchema.methods.getLeadboard = function () {};

const Wall = mongoose.model("Wall", wallSchema);

// --- INDOOR WALL ---
const indoorSchema = new mongoose.Schema(
    {
        facility: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Facility",
            required: true,
        },
    },
    {
        toJSON: {
            transform: function (doc, ret) {
                baseWallTransform(doc, ret);
                return ret;
            },
        },
    },
);

indoorSchema.methods.editWall = function (updates) {};

const IndoorWall = Wall.discriminator("IndoorWall", indoorSchema);

// --- OUTDOOR WALL ---
const outdoorSchema = new mongoose.Schema(
    {
        publicBody: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "PublicBody",
            required: true,
        },
    },
    {
        toJSON: {
            transform: function (doc, ret) {
                baseWallTransform(doc, ret);
                return ret;
            },
        },
    },
);

outdoorSchema.methods.editWall = function (updates) {};

const OutdoorWall = Wall.discriminator("OutdoorWall", outdoorSchema);

module.exports = {
    Wall,
    IndoorWall,
    OutdoorWall,
};

File: frontend/lib/models/wall.dart
import 'climbing_session.dart';

class Wall {
  final String id;
  final String name;
  final double latitude;
  final double longitude;
  final String description;
  final String difficulty;
  final String wallType;
  final String? ownerName;
  final List<ClimbingSession> sessions;
  final double rating;
  final int totalSessions;

  Wall({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.description,
    required this.difficulty,
    required this.wallType,
    this.ownerName,
    required this.sessions,
    this.rating = 0.0,
    this.totalSessions = 0,
  });

  // Compatibility with older UI code that expects `type`.
  String get type => wallType;

  factory Wall.fromJson(Map<String, dynamic> json) {
    final coordinates = json['location']?['coordinates'];
    double latitude = _parseDouble(json['latitude']);
    double longitude = _parseDouble(json['longitude']);

    if (coordinates is List && coordinates.length >= 2) {
      longitude = _parseDouble(coordinates[0]);
      latitude = _parseDouble(coordinates[1]);
    }

    final wallType = (json['wallType'] ?? json['type'] ?? 'Wall').toString();

    String? owner;
    if (json['facility'] is Map) {
      owner = json['facility']['username']?.toString();
    } else if (json['publicBody'] is Map) {
      owner = json['publicBody']['username']?.toString();
    }

    final sessionsRaw = json['sessions'];
    final sessions = <ClimbingSession>[];
    if (sessionsRaw is List) {
      for (final item in sessionsRaw) {
        if (item is Map<String, dynamic>) {
          sessions.add(ClimbingSession.fromJson(item));
        } else if (item is Map) {
          sessions.add(ClimbingSession.fromJson(Map<String, dynamic>.from(item)));
        }
      }
    }

    double parseRating(dynamic r) {
      if (r == null) return 0.0;
      if (r is num) return r.toDouble();
      if (r is String) return double.tryParse(r) ?? 0.0;
      return 0.0;
    }

    int parseTotalSessions(dynamic s) {
      if (s == null) return 0;
      if (s is int) return s;
      if (s is num) return s.toInt();
      if (s is String) return int.tryParse(s) ?? 0;
      return 0;
    }

    final rating = parseRating(json['rating']);
    final totalSessions = parseTotalSessions(json['totalSessions'] ?? json['total_sessions']);

    return Wall(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      name: (json['name'] ?? 'Unknown Wall').toString(),
      latitude: latitude,
      longitude: longitude,
      description: (json['description'] ?? 'No description available.').toString(),
      difficulty: (json['difficulty'] ?? 'UNKNOWN').toString(),
      wallType: wallType,
      ownerName: owner,
      sessions: sessions,
      rating: rating,
      totalSessions: totalSessions,
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'latitude': latitude,
      'longitude': longitude,
      'description': description,
      'difficulty': difficulty,
      'wallType': wallType,
      'ownerName': ownerName,
      'sessions': sessions.map((session) => session.toJson()).toList(),
      'rating': rating,
      'totalSessions': totalSessions,
    };
  }
}
